Add support for InfluxDB 0.9

peggyl opened the pull request:
This change is based on the patch created in https://github.com/python-diamond/Diamond/pull/148 (with a small fix to an incorrect key name).

The following changes should be made in the configuration file (default location: `/etc/diamond/diamond.conf`) to support InfluxDB 0.9:

```
handlers = diamond.handler.influxdbHandler.InfluxdbHandler # plus other handlers used

...

[[InfluxdbHandler]]
# note the version defaults to 0.8 in the Diamond code, but that could be changed to 0.9
influxdb_version = 0.9 
# others: hostname, port, username, password, database, time_precision, batch_size, cache_size
```

This was tested using `influxdb==2.9.1` and `Diamond==4.0.193` (installed via pip).

**Testing**: successfully sent Diamond-collected metrics from my VM to the staging InfluxDB.

peggyl commented:
HOLD: scanning logs and still need to resolve this error around float/int compatibility:

```
InfluxDBClientError: 400: write failed: field type conflict: input field "value" on measurement "servers.vagrant-ubunt
u-trusty-64.cpu.cpu1.user" is type float64, already exists as type integer
```

nathanleiby commented:
Sadly it looks like there aren't existing tests for this handler to build upon :/ https://github.com/Clever/Diamond/tree/INFRA-1066-influx-0.9/src/diamond/handler/test

nathanleiby commented:
LGTM barring your current investigation

peggyl commented:
**Context**: We initially sent 0i for some metrics, so they were stored as ints. Later, when we tried to send floats (e.g. 0.5) for the same measurement, InfluxDB would reject them: https://github.com/influxdb/influxdb/blob/2c4622c95432b7d7e863231fa69c99997dcab973/tsdb/shard.go#L413

This is a known issue, e.g. https://github.com/influxdb/influxdb/issues/3032. A quick fix is to have the client send all numeric values as floats and never as ints. We can do this by not appending 'i' to int values (see 0.9.3 changes in https://github.com/influxdb/influxdb/blob/master/CHANGELOG.md).

This works when testing in staging, but unfortunately required dropping the measurements that were typed as ints before new float values could be successfully added. We may need some data migration work in production; I'm looking into this.

Also, this fix is most easily made in influxdb-python by skipping https://github.com/influxdb/influxdb-python/blob/master/influxdb/line_protocol.py#L62. We will need a fork of this, or add preprocessing in the Diamond handler (convert all ints to floats before sending to the InfluxDB client). Thoughts?

nathanleiby commented:
@peggyl Sounds like Influx is generally happy treating all numbers as floats unless explicitly specified otherwise: https://github.com/influxdb/influxdb/issues/3032#issuecomment-128511512

I think that should be fine for our Diamond usecase. I think I would cast ints to floats within the Diamond InfluxHandler, and document it clearly, rather than changing both libraries. Not a strong opinion though.

I'm not yet sure of our data migration plan -- do we have a way to easily move data from our current influx DB to influx9 DB? Or will it there be a separation when we move from current to new system? cc: @mohit @dfuentes

peggyl commented:
@nathanleiby https://github.com/Clever/Diamond/commit/f9b3be85fca3659b9660118d231c1f164ab989fb casts any int values to floats before sending them to the InfluxDB client, which is what actually checks if a value is an int (and if so, appends 'i') and posts measurements to the Influx server. We don't currently have a fork of that repo and this should be simple enough to do here, in Diamond's InfluxDB handler. :)

mohit commented:
@peggyl quick question. Can we run both influx0.8 and influx0.9 outputs together.

mohit commented on src/diamond/handler/influxdbHandler.py:
```diff
@@ -170,11 +175,24 @@ def _send(self):
+                                "columns": ["time", "value"]})
+                    elif self.influxdb_version == "0.9":
+                        for path in self.batch:
+                            # HACK: convert int to float before writing to InfluxDB
```
please link the appropriate PR/issue on the influxdb repo here so that we can track progress.

peggyl replied:
I don't think we're waiting on any changes in the influxdb repo.

According to https://github.com/influxdb/influxdb/issues/3032, ints and floats are handled correctly server-side and it's up to the client to send the right data type (e.g. if the client sends "2i", it will be treated as an int, while "2" or "2.0" is a float). 

The client handles this in the [influxdb-python repo](https://github.com/influxdb/influxdb-python/blob/master/influxdb/line_protocol.py#L62), but since we don't have a Clever/influxdb-python fork, I thought it'd be easier to add a quick preprocessing step here. I don't expect this will be changed in influxdb-python since it was just changed in 0.9.3.

mohit replied:
:+1:

mohit commented:
@peggyl before we merge this, we would want to build some dashboards based on the data to ensure that everything is working well.

mohit commented on src/diamond/handler/influxdbHandler.py:
```diff
@@ -29,8 +29,10 @@
 password = root
 database = graphite
 time_precision = s
+influxdb_version = 0.8
```
should be a string?

peggyl replied:
Okay for it to be a string in the conf file, since all values are parsed as strings by default.

mohit commented on src/diamond/handler/influxdbHandler.py:
```diff
@@ -120,6 +124,7 @@ def get_default_config(self):
             'batch_size': 1,
             'cache_size': 20000,
             'time_precision': 's',
+            'influxdb_version': 0.8,
```
use string?

peggyl replied:
Good catch, thanks!

mohit commented on src/diamond/handler/influxdbHandler.py:
```diff
@@ -170,11 +175,24 @@ def _send(self):
-                            "points": self.batch[path],
-                            "name": path,
-                            "columns": ["time", "value"]})
+                    if self.influxdb_version == "0.8":
```
since this is a string ^

mohit commented:
@peggyl assigning to you for actions on comments :smile_cat:

peggyl commented:
@mohit responded to most comments inline. In terms of running 0.8 and 0.9 together – it looks like the schema has breaking changes, so we probably can't send the same output to both servers. We could tweak the config files and code to send each one the correctly formatted data for its version?

I think some work's been done on export functionality for 0.8 -> 0.9: https://github.com/influxdb/influxdb/pull/3477

peggyl commented:
@nathanleiby assigning to you to review changes. I'll take a look at adding InfluxdbHandler test coverage and talk with @mohit about dashboard testing tomorrow.

nathanleiby commented:
LGTM

nathanleiby commented:
@peggyl this looks good - please feel free to schedule some time with me to deploy in staging / roll out generally

peggyl commented:
@nathanleiby Latest commit should enable writing to both InfluxDB 0.8 (production) and 0.9 (staging). I believe we just remove the Influxdb9Handler and update the main conf once we're just using 0.9 in prod.

nathanleiby commented:
@peggyl that sounds good! let's get this running on some machines so we can get more real data into Influx9 / start rebuilding grafana graphs with new schema

peggyl commented:
@nathanleiby LGTY to rebase and merge here? I'll need the commit hash on master (ideally) before updating salt.

nathanleiby commented:
:+1:

peggyl committed:
Update InfluxdbHandler to support InfluxDB 0.9.

This also creates a second handler, Influxdb9Handler,
to facilitate writing to two Influx DBs by making changes
only to Diamond handler configuration, no code changes.
